lib/ptracer.h

```diff
@@ -0,0 +1,485 @@
+               global table. */
+            if(!strcmp(lua_tostring(L, -2), "_G")) {
+                /* Remove value and continue. */
+                lua_pop(L, 1);
+                continue;
+            }
+
+            /* Is it the function we are looking for? */
+            if(lua_rawequal(L, fn_idx, -1)) {
+                /* Remove value and keep name. */
+                lua_pop(L, 1);
+                return true;
+            }
+            /* If not go one level deeper and get the value recursively. */
+            else if(_pallene_tracer_findfield(L, fn_idx, level - 1)) {
+                /* Remove the table but keep name. */
+                lua_remove(L, -2);
+
+                /* Add a "." in between. */
+                lua_pushliteral(L, ".");
+                lua_insert(L, -2);
+
+                /* Concatenate last 3 values, resulting "table.some_func". */
+                lua_concat(L, 3);
+
+                return true;
+            }
+        }
+
+        /* Pop the value. */
+        lua_pop(L, 1);
+    }
+
+    return false;
+}
+
+/* Pushes a function name if found in the global table and returns true.
+   Returns false otherwise. */
+/* Expects the function to be pushed in the stack. */
+static bool _pallene_tracer_pgf_name(lua_State *L) {
+    int top = lua_gettop(L);
+
+    /* Start from the global table. */
+    lua_pushglobaltable(L);
+
+    if(_pallene_tracer_findfield(L, top, 2)) {
+        lua_remove(L, -2);
+        return true;
+    }
+
+    lua_pop(L, 1);
+    return false;
+}
+
+/* Returns the maximum number of levels in Lua stack. */
+static int _pallene_tracer_countlevels(lua_State *L) {
+    lua_Debug ar;
+    int li = 1, le = 1;
+
+    /* Find an upper bound */
+    while (lua_getstack(L, le, &ar)) {
+        li = le, le *= 2;
+    }
+
+    /* Do a binary search */
+    while (li < le) {
+        int m = (li + le) / 2;
+
+        if (lua_getstack(L, m, &ar)) li = m + 1;
+        else le = m;
+    }
+
+    return le - 1;
+}
+
+/* Counts the number of white and black frames in the Pallene call stack. */
+static void _pallene_tracer_countframes(pt_cont_t *cont, int *mwhite, int *mblack) {
+    *mwhite = *mblack = 0;
+
+    for(int i = 0; i < cont->count; i++) {
+        *mwhite += (cont->stack[i].type == PALLENE_TRACER_FRAME_TYPE_C);
+        *mblack += (cont->stack[i].type == PALLENE_TRACER_FRAME_TYPE_LUA);
+    }
+}
+
+/* Responsible for printing and controlling some of the traceback fn parameters. */
+static void _pallene_tracer_dbg_print(const char *buf, bool *ellipsis, int *pframes, int nframes) {
+    /* We have printed the frame, even tho it might not be visible ;). */
+    (*pframes)++;
+
+    /* Should we print? Are we at any point in top or bottom printing threshold? */
+    bool should_print = (*pframes <= PALLENE_TRACEBACK_TOP_THRESHOLD)
+        || ((nframes - *pframes) <= PALLENE_TRACEBACK_BOTTOM_THRESHOLD);
+
+    if(luai_likely(should_print == true))
```

Comment:
```suggestion
    if(luai_likely(should_print))
```

Reply:
I do not know how to use `luai_(un)likely` macros correctly to be honest. I've been using `__builtin_expect` differently than defined in `luai_(un)likely` macros.